simulator/host_datastore_system.go

```diff
@@ -60,11 +60,23 @@ func (dss *HostDatastoreSystem) add(ctx *Context, ds *Datastore) *soap.Fault {
 	}
 
 	folder := ctx.Map.getEntityFolder(dss.Host, "datastore")
-	ds.Self.Type = typeName(ds)
-	// Datastore is the only type where create methods do not include the parent (Folder in this case),
-	// but we need the moref to be unique per DC/datastoreFolder, but not per-HostSystem.
-	ds.Self.Value += "@" + folder.Self.Value
-	// TODO: name should be made unique in the case of Local ds type
+
+	found := false
+	if e := ctx.Map.FindByName(ds.Name, folder.ChildEntity); e != nil {
+		if e.Reference().Type != "Datastore" {
+			return Fault(e.Reference().Value, &types.DuplicateName{
+				Name:   ds.Name,
+				Object: e.Reference(),
+			})
+		}
+
+		// if datastore already exists, use current reference
+		found = true
+		ds.Self = e.Reference()
+	} else {
+		// put datastore to folder and generate reference
+		folderPutChild(ctx, folder, ds)
```

Comment:
generated reference value inside `folderPutChild` is unique per DC/datastoreFolder.

https://github.com/vmware/govmomi/blob/dbc80bac02c950760bc1e4d5e7cddbc61ce4e524/simulator/registry.go#L144